tools: Add nispor-in-container

ihuguet opened the pull request:
Script to create a container where nispor can easily be tested. It has some small limitations, like not being able to load kernel modules or creating netdevsim devices. The first problem can be workarounded loading the modules in the host before. I haven't found any solution for the second, so far.

ihuguet committed:
tools: Add nispor-in-container

Script to create a container where nispor can easily be tested. It has
some small limitations, like not being able to load kernel modules or
creating netdevsim devices. The first problem can be workarounded
loading the modules in the host before. I haven't found any solution for
the second, so far.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

cathay4t commented:
This PR just share a personal script to upstream. Please provide information why it will be used by others.

Considering nispor is a single binary, anyone can build a container using it simply.

ihuguet commented:
The container created by the script has the packages required to build and test nispor inside the container already installed, and it mounts the root folder of the project so you can easily build and install.

It is a convenient way to build the project and run the tests locally for new contributors. Maintainers and long standing contributors might have their personal and customized scripts, containers, etc, but people that only intends to do a few contributions won't have them, and they are still required to run the tests before submitting a PR.

Indeed, you can manually build a container but not everyone knows well how to do it. If you also want to mount the project and build it inside the container, you will need to manually install the dependencies, investigating which ones they are first. Again, this is very low effort for long standing contributors, but it might be tedious for sporadic contributors, though discouraging them from contributing.

I even suggest to add it to DEVEL.md so new contributors know of its existence.

So, TLDR: this is convenient to facilitate the workflow, specially to sporadic contributors (but not only).

cathay4t commented:
`podman run -it --rm fedora bash` and run `make check` there is not hard.

I do not want to maintain a script I do not use.

And I use VM with source code mounted via NFS instead of container.

Developer should have their own stuff to make work done.

ihuguet commented:
> This PR just share a personal script to upstream

It's not a personal script, I've written it with sporadic contributors in mind. I wouldn't try to merge a script that only fits my personal tastes.

> `podman run -it --rm fedora bash` and run `make check` there is not hard. 

It won't work, you also need to mount the project's folder and install the dependencies. Still, you might think that this is easy too, but not everyone uses containers often, so they might need to invest time in researching how to do it. Then, they will need to do trial and error to see what dependencies do they need in order to build inside.

Anyone would be OK to invest that time if he's going to contribute frequently, but it is too much for a single small contribution. 

> I do not want to maintain a script I do not use.

You don't need to maintain everything, there are more contributors. You don't need to use it if you have prepared a better environments, but as I say, this is targeted to sporadic contributors (and also useful for frequent contributors if they want).

Also, it is very simple, so easy to maintain. And as it is not critical for the workflow, fixing it if it breaks wouldn't be urgent, you can leave the fix for those who use it or for other contributors, like me.

> Developer should have their own stuff to make work done.

To submit a PR you are supposed to run the tests locally first, it is a required step, and they mess your system if you run them without a container or VM. Facilitating this required step to new contributors doesn't sound as a bad idea to me. It's not a weird use case that only a few developers can take advantage of.

I agree that we don't have to add "personal scripts", but I think that this covers a gap in the project and can be useful for almost everyone.

ffmancera commented:
I have been reviewing the script. I think it would be nice to have it on the project. While Gris comment mentioning it can be done manually it is true, it is also true that not everyone would be willing to do that for a casual contribution or even if they are willing to do it, they might find problems.

When I developed something on Nispor I usually used a similar script from another project (NetworkManager) and adapted it to Nispor. So, having this directly would be nice.

+1 from my side.

cathay4t commented:
Regardless my reluctance, please expand `Makefile` to use this script as `make check_in_container` and also update README.md to explain the usage of it.

Please also expand Ci to test this script also, I don't want someone ask me about why this script is not working anymore.

ihuguet committed:
nispor-in-container: allow specifying a command

It already allowed to specify arguments to pass to `podman run`.
Allow to specify also a command to run. To differentiate arguments from
commands we need to use the `--` separator.

This allows to run, for example: tools/nispor-in-container -- make check

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet committed:
Makefile: add check_in_container

Add make target to run the tests in a container using
nispor-in-container.

Add to the documentation this method to test, and also running
nispor-in-container directly.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet committed:
nispor-in-container: fallback to Docker if Podman is missing

Not tested with Docker, actually, but it accepts all the arguments that
we are passing here to docker-build and docker-run, so it should work
fine.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet commented:
Done, I have added a `make check_in_container` target and added explanations to `DEVEL.md`.

I have also added a fallback to use Docker if Podman is missing.

ihuguet committed:
CI: add test to test that nispor-in-container works

Note that there is at least one test that fails within the container
because netdevsim devices cannot be created from inside. This means that
we cannot run `make check` in the CI because it will always fail unless
we fix that test.

Instead, just check that the container is built and run succesfully.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet commented:
Added a CI test too, I have missed that request.

liangwen12year commented on tools/nispor-in-container:
```diff
@@ -3,10 +3,20 @@
+    if command -v podman &>/dev/null; then
+        PODMAN=podman
+    elif command -v docker &>/dev/null; then
+        PODMAN=docker  # why not?
```
Anticipatedly, the same arg works for podman and should also work for docker.

liangwen12year commented:
I have tested it locally, I did not find any problem.

cathay4t commented on .github/workflows/main.yml:
```diff
@@ -70,3 +70,12 @@ jobs:
+    - uses: actions/checkout@v4
+
+    - name: Test nispor-in-container
+      run: tools/nispor-in-container -- echo success
```
This CI just invoked a container and do nothing but echo.

I guess `make check_in_container` should be used here.

ihuguet replied:
I knew there were a reason why I didn't do that way: inside the container, there are a few tests that fails because netdevsim cannot be configured from the container. So we cannot do `make check_in_container` in the CI. It is OK if I revert the last change and leave `tools/nispor-in-container -- echo success`? That will check that nispor-in-container script is OK.

cathay4t replied:
A simple echo cannot prove it is working.

Please use
```
    - name: Install extra kernel module(like netdevsim)
      run: |
        sudo apt-get update;
        sudo apt-get -y install "linux-modules-extra-$(uname -r)"

```

ihuguet replied:
That's not enough. The container is still a rootless container and cannot configure netdevsim.

cathay4t commented:
This PR has stale for a while.

I still see no point of share personal script to upstream without CI verifying it.

Move to close it on 2024 Aug 31.

cathay4t commented:
Closing as stale.
